quantum/process_keycode/process_auto_shift.c

```diff
@@ -107,114 +92,81 @@ static bool autoshift_press(uint16_t keycode, keyrecord_t *record) {
- *
- * If another key is pressed, register the key.
  */
-static void autoshift_check_record(uint16_t keycode, keyrecord_t *record) {
-    if (!autoshift_flags.in_progress) {
-        return;
-    }
+static void autoshift_end(uint16_t keycode, uint16_t now, bool matrix_trigger) {
+    // Called on key down with KC_NO, auto-shifted key up, and timeout.
+    if (autoshift_flags.in_progress) {
+        // Process the auto-shiftable key.
+        autoshift_flags.in_progress = false;
 
-    // Process the release of the autoshiftable key.
-    if (keycode == autoshift_lastkey && !record->event.pressed) {
-        autoshift_flush_shift();
         // Time since the initial press was recorded.
-        const uint16_t elapsed = TIMER_DIFF_16(record->event.time, autoshift_time);
-        if (!autoshift_flags.registered && elapsed < autoshift_timeout) {
-            // Auto-shiftable key is being released before the shift timeout;
-            // simulate the original press then let the usual processing take
-            // care of the release.
-            register_code(keycode);
+        const uint16_t elapsed = TIMER_DIFF_16(now, autoshift_time);
+        if (elapsed < autoshift_timeout) {
+            register_code(autoshift_lastkey);
             autoshift_flags.lastshifted = false;
-            autoshift_flags.registered  = true;
-#    if TAP_CODE_DELAY > 0
-            wait_ms(TAP_CODE_DELAY);
+        } else {
+            // Simulate pressing the shift key.
+            add_weak_mods(MOD_BIT(KC_LSFT));
+            register_code(autoshift_lastkey);
+            autoshift_flags.lastshifted = true;
+#    if defined(AUTO_SHIFT_REPEAT) && !defined(AUTO_SHIFT_NO_AUTO_REPEAT)
+            if (matrix_trigger) {
+                // Prevents release.
+                return;
+            }
 #    endif
         }
-        autoshift_flush(record->event.time);
-    } else if (record->event.pressed) {
-        if (keycode == KC_LSFT) {
-            // If the user presses shift, auto-shift will not hold shift for
-            // them.
-            autoshift_flags.holding_shift = false;
-        } else if (!autoshift_flags.registered) {
-            // If the key isn't registered yet, it means the timeout hasn't
-            // elapsed, so register the key without additional shifting.
-            autoshift_flush(0);
-            autoshift_lastkey = KC_NO;
-        } else {
-            // The key is registered; flush any shift state for the
-            // non-autoshiftable key.
-            autoshift_flush_shift();
+
+#    if TAP_CODE_DELAY > 0
+        wait_ms(TAP_CODE_DELAY);
+#    endif
+        unregister_code(autoshift_lastkey);
+        del_weak_mods(MOD_BIT(KC_LSFT));
+    } else {
+        // Release after keyrepeat.
+        unregister_code(keycode);
+        if (keycode == autoshift_lastkey) {
+            // This will only fire when the key was the last auto-shiftable
+            // pressed. That prevents aaaaBBBB then releasing a from unshifting
```

Comment:
nit: do you mind putting quotes around the keys here ("releasing a from" is a bit hard to parse eagerly).